llama_index/llms/openai.py

```diff
@@ -247,71 +290,26 @@ def gen() -> ChatResponseGen:
                 else:
                     delta = {}
 
-                if response_id is None:
-                    response_id = response.id
-
-                if response_id != response.id:
-                    # start of a new response
-                    is_function = False
-                    response_ix = 0
+                # check if this chunk is the start of a function call
+                if (delta.role == MessageRole.ASSISTANT) and (delta.content is None):
+                    is_function = True
 
                 # update using deltas
                 role = delta.role or MessageRole.ASSISTANT
                 content_delta = delta.content or ""
                 content += content_delta
 
-                # TODO: REFACTOR WITH FUNCTIONS
-                # openai provides the delta on one tool-at-a-time
```

Comment:
@[jerryjliu](https://github.com/jerryjliu) jerryjliu [yesterday](https://github.com/run-llama/llama_index/pull/8738#discussion_r1386005333)
high-level this is a lot of nested code, can we abstract out into helper functions above?

Member
@[jerryjliu](https://github.com/jerryjliu) jerryjliu [yesterday](https://github.com/run-llama/llama_index/pull/8738#discussion_r1386005505)
same with the async version